src/analysis-cache.ts

```diff
@@ -0,0 +1,169 @@
+    const newCache = this._clone();
+    newCache.parsedDocumentPromises.delete(path);
+    newCache.scannedDocumentPromises.delete(path);
+    newCache.dependenciesScanned.delete(path);
+    newCache.scannedDocuments.delete(path);
+    newCache.analyzedDocuments.delete(path);
+
+    // Analyzed documents need to be treated more carefully, because they have
+    // relationships with other documents. So first we remove all documents
+    // which transitively import the changed document. We also need to mark
+    // all of those docs as needing to rescan their dependencies.
+    for (const invalidatedPath of dependants) {
+      newCache.dependenciesScanned.delete(invalidatedPath);
+      newCache.analyzedDocuments.delete(invalidatedPath);
+    }
+    // Then we clear out the analyzed document promises, which could have
+    // in-progress results that don't cohere with the state of the new cache.
+    // Only populate the new analyzed promise cache with results that are
+    // definite, and not invalidated.
+    newCache.analyzedDocumentPromises.clear();
+    for (const keyValue of newCache.analyzedDocuments) {
+      newCache.analyzedDocumentPromises.set(
+          keyValue[0], Promise.resolve(keyValue[1]));
+    }
+
+    return newCache;
+  }
+
+  private _clone(): AnalysisCache {
+    const newCache = new AnalysisCache();
+    this._copyMapInto(
+        this.parsedDocumentPromises, newCache.parsedDocumentPromises);
+    this._copyMapInto(
+        this.scannedDocumentPromises, newCache.scannedDocumentPromises);
+    this._copyMapInto(
+        this.analyzedDocumentPromises, newCache.analyzedDocumentPromises);
+    this._copyMapInto(this.dependenciesScanned, newCache.dependenciesScanned);
+    this._copyMapInto(this.scannedDocuments, newCache.scannedDocuments);
+    this._copyMapInto(this.analyzedDocuments, newCache.analyzedDocuments);
+    return newCache;
+  }
+
+  private _copyMapInto<K, V>(from: Map<K, V>, into: Map<K, V>) {
+    for (const kv of from) {
+      into.set(kv[0], kv[1]);
+    }
+  }
+}
+
+/**
+ * Trawl the import graph and return the paths of (transitive) dependants on
+ * the given path.
+ *
+ * This should eventually live somewhere like Document, but we really need
+ * something like a Project, which encapsulates all of the known Documents
+ * inside a basedir, and which has the same interface as Document. That is the
+ * object that the AnalysisCache needs to know the results of calling
+ * getImportersOf.
+ */
+export function getImportersOf(
+    path: string,
+    documents: Iterable<Document>,
+    scannedDocuments: Iterable<ScannedDocument>,
+    urlResolver: (url: string) => string): Set<string> {
+  const invertedIndex =
+      _buildInvertedIndex(documents, scannedDocuments, urlResolver);
+  const visited = new Set<string>();
+  const toVisit = new Set<string>([path]);
+  while (toVisit.size > 0) {
+    const path = toVisit.keys().next().value!;
```

Comment:
interesting... Can you use `entries()` for clarify? `keys()` on `Set` is odd to read